src/sentry/integrations/gitlab/webhooks.py

```diff
@@ -79,92 +142,79 @@ def create_commits(self, event, organization, repo):
                         ).astimezone(timezone.utc),
                     )
-
             except IntegrityError:
                 pass
 
 
-class MergeEventWebhook(Webhook):
-    def __call__(self, organization, event):
-        repo = self.create_repo(event, organization)
-        self.create_commits(event, organization, repo)
-
-
-class PushEventWebhook(Webhook):
-    def __call__(self, organization, event):
-        repo = self.create_repo(event, organization)
-        self.create_commits(event, organization, repo)
-
-
 class GitlabWebhookEndpoint(View):
+    provider = 'gitlab'
+
     _handlers = {
-        'repository_update': PushEventWebhook,
-        'merge_request': MergeEventWebhook,
+        'Push Hook': PushEventWebhook,
+        'Merge Request Hook': MergeEventWebhook,
     }
 
-    def get_handler(self, event_type):
-        return self._handlers.get(event_type)
-
     @method_decorator(csrf_exempt)
     def dispatch(self, request, *args, **kwargs):
         if request.method != 'POST':
             return HttpResponse(status=405)
 
-        # if not self.check_secret(request):
-        #     raise Exception???
-
         return super(GitlabWebhookEndpoint, self).dispatch(request, *args, **kwargs)
 
-    def post(self, request, organization_id):
+    def post(self, request):
         try:
-            organization = Organization.objects.get_from_cache(
-                id=organization_id,
-            )
-        except Organization.DoesNotExist:
-            logger.error(
-                PROVIDER_NAME + '.webhook.invalid-organization',
+            # Munge the token to extract the integration external_id.
+            # gitlab hook payloads don't give us enough unique context
+            # to find data on our side so we embed one in the token.
+            token = request.META['HTTP_X_GITLAB_TOKEN']
+            instance, group_path, secret = token.split(':')
+            external_id = u'{}:{}'.format(instance, group_path)
+        except Exception:
+            logger.info('gitlab.webhook.invalid-token', extra={'token': token})
+            return HttpResponse(status=400)
+
+        try:
+            integration = Integration.objects.filter(
+                provider=self.provider,
+                external_id=external_id
+            ).prefetch_related('organizations').get()
+        except Integration.DoesNotExist:
+            logger.info(
+                'gitlab.webhook.invalid-organization',
                 extra={
-                    'organization_id': organization_id,
+                    'external_id': request.META['HTTP_X_GITLAB_TOKEN'],
                 }
             )
             return HttpResponse(status=400)
 
-        body = six.binary_type(request.body)
-        if not body:
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-body', extra={
-                    'organization_id': organization.id,
+        if not constant_time_compare(secret, integration.metadata['webhook_secret']):
```

Comment:
This is cool. I hadn't seen this before. Maybe we could add this to the others as well?

Reply:
Yes. I'll add an issue in jira.